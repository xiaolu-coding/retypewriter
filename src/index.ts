import type { Diff } from 'diff-match-patch'
import { diff_match_patch as DMP } from 'diff-match-patch'

export interface InsertPatch {
  type: 'insert'
  from: number
  text: string
}

export interface RemovalPatch {
  type: 'removal'
  from: number
  length: number
}

// 集合类型
export type Patch = InsertPatch | RemovalPatch

// 将增加修改转换为 patches对象数组，根据type属性得知是增加还是删除
export function calculatePatch(diff: Diff[]): Patch[] {
  const patches: Patch[] = []
  // 指针，记录位置
  let index = 0
  for (const change of diff) {
    // change[0] -1是删除， 0是没变，1是增加
    if (change[0] === 0) {
      // 如果没变，指针移动到这句话的最后
      index += change[1].length
      continue
    }
    else if (change[0] === -1) {
      const length = change[1].length
      patches.push({
        type: 'removal',
        from: index + length, // 删除要加上删除的长度
        length,
      })
    }
    else if (change[0] === 1) {
      patches.push({
        type: 'insert',
        from: index,
        text: change[1],
      })
      // 加了之后，指针要移动
      index += change[1].length
    }
    else {
      throw new Error('error')
    }
  }
  return patches
}
// 通过遍历patches的对象数组，执行相应的操作，得到最终结果
export function applyPatches(input: string, patches: Patch[]) {
  let output = input
  for (const patch of patches) {
    if (patch.type === 'insert') {
      // 将两边取出来然后加进去
      output = output.slice(0, patch.from) + patch.text + output.slice(patch.from)
    }
    else if (patch.type === 'removal') {
      // 取出要删除的前面的和后面的加起来
      output = output.slice(0, patch.from - patch.length) + output.slice(patch.from)
    }
  }
  return { output }
}

// diff对比文本 返回change数组，数组内部也是数组，索引0为 -1 0 1，索引1为相应内容
export function diff(a: string, b: string): Diff[] {
  const differ = new DMP()
  const delta = differ.diff_main(a, b)
  differ.diff_cleanupSemantic(delta)
  return delta
}
