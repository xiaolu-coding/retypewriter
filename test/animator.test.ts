import { expect, it } from 'vitest'
import { calculatePatch, createAnimator, diff } from '../src'
import { input, output } from './fixture'

it('animator', () => {
  const delta = diff(input, output)
  const patches = calculatePatch(delta)
  const animator = createAnimator(input, patches)
  expect([...animator]).toMatchSnapshot('animator')
})
