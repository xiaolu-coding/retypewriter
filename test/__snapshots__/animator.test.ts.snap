// Vitest Snapshot v1

exports[`animator > animator 1`] = `
[
  {
    "cursor": 75,
    "output": "
import { describe, expect, it } from 'vitest'
import { one } from '../src'

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 74,
    "output": "
import { describe, expect, it } from 'vitest'
import { one } from '../src

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 73,
    "output": "
import { describe, expect, it } from 'vitest'
import { one } from '../sr

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 72,
    "output": "
import { describe, expect, it } from 'vitest'
import { one } from '../s

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 71,
    "output": "
import { describe, expect, it } from 'vitest'
import { one } from '../

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 70,
    "output": "
import { describe, expect, it } from 'vitest'
import { one } from '..

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 69,
    "output": "
import { describe, expect, it } from 'vitest'
import { one } from '.

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 68,
    "output": "
import { describe, expect, it } from 'vitest'
import { one } from '

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 67,
    "output": "
import { describe, expect, it } from 'vitest'
import { one } from 

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 66,
    "output": "
import { describe, expect, it } from 'vitest'
import { one } from

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 65,
    "output": "
import { describe, expect, it } from 'vitest'
import { one } fro

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 64,
    "output": "
import { describe, expect, it } from 'vitest'
import { one } fr

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 63,
    "output": "
import { describe, expect, it } from 'vitest'
import { one } f

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 62,
    "output": "
import { describe, expect, it } from 'vitest'
import { one } 

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 61,
    "output": "
import { describe, expect, it } from 'vitest'
import { one }

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 60,
    "output": "
import { describe, expect, it } from 'vitest'
import { one 

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 59,
    "output": "
import { describe, expect, it } from 'vitest'
import { one

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 58,
    "output": "
import { describe, expect, it } from 'vitest'
import { on

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 57,
    "output": "
import { describe, expect, it } from 'vitest'
import { o

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 56,
    "output": "
import { describe, expect, it } from 'vitest'
import { 

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 55,
    "output": "
import { describe, expect, it } from 'vitest'
import {

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 54,
    "output": "
import { describe, expect, it } from 'vitest'
import 

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 53,
    "output": "
import { describe, expect, it } from 'vitest'
import

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 52,
    "output": "
import { describe, expect, it } from 'vitest'
impor

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 51,
    "output": "
import { describe, expect, it } from 'vitest'
impo

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 50,
    "output": "
import { describe, expect, it } from 'vitest'
imp

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 49,
    "output": "
import { describe, expect, it } from 'vitest'
im

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 48,
    "output": "
import { describe, expect, it } from 'vitest'
i

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 47,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 90,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 89,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('exporte', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 88,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('export', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 87,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('expor', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 86,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('expo', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 85,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('exp', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 84,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('ex', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 83,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('e', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 82,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 83,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('o', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 84,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('on', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 85,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 124,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
   })
})
",
  },
  {
    "cursor": 125,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    })
})
",
  },
  {
    "cursor": 126,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
     })
})
",
  },
  {
    "cursor": 127,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
      })
})
",
  },
  {
    "cursor": 128,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    e  })
})
",
  },
  {
    "cursor": 129,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    ex  })
})
",
  },
  {
    "cursor": 130,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    exp  })
})
",
  },
  {
    "cursor": 131,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expe  })
})
",
  },
  {
    "cursor": 132,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expec  })
})
",
  },
  {
    "cursor": 133,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect  })
})
",
  },
  {
    "cursor": 134,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(  })
})
",
  },
  {
    "cursor": 135,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(t  })
})
",
  },
  {
    "cursor": 136,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(tw  })
})
",
  },
  {
    "cursor": 137,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two  })
})
",
  },
  {
    "cursor": 138,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two)  })
})
",
  },
  {
    "cursor": 139,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).  })
})
",
  },
  {
    "cursor": 140,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).t  })
})
",
  },
  {
    "cursor": 141,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).to  })
})
",
  },
  {
    "cursor": 142,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toE  })
})
",
  },
  {
    "cursor": 143,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEq  })
})
",
  },
  {
    "cursor": 144,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqu  })
})
",
  },
  {
    "cursor": 145,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqua  })
})
",
  },
  {
    "cursor": 146,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual  })
})
",
  },
  {
    "cursor": 147,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(  })
})
",
  },
  {
    "cursor": 148,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2  })
})
",
  },
  {
    "cursor": 149,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)  })
})
",
  },
  {
    "cursor": 150,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
]
`;
